pkg/controllermanager/controller/credentialsbinding/reconciler.go

```diff
@@ -0,0 +1,258 @@
+		log.Info("Adding finalizer to secret", "secret", client.ObjectKeyFromObject(secret))
+		if err := controllerutils.AddFinalizers(ctx, r.Client, secret, gardencorev1beta1.ExternalGardenerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not add finalizer to secret: %w", err)
+		}
+	}
+
+	labelKey := v1beta1constants.LabelShootProviderPrefix + credentialsBinding.Provider.Type
+	if !metav1.HasLabel(secret.ObjectMeta, labelKey) {
+		patch := client.MergeFrom(secret.DeepCopy())
+		metav1.SetMetaDataLabel(&secret.ObjectMeta, labelKey, "true")
+		if err := r.Client.Patch(ctx, secret, patch); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add provider type label to Secret referenced in CredentialsBinding: %w", err)
+		}
+	}
+
+	if len(credentialsBinding.Quotas) != 0 {
+		for _, objRef := range credentialsBinding.Quotas {
+			quota := &gardencorev1beta1.Quota{}
+			if err := r.Client.Get(ctx, client.ObjectKey{Namespace: objRef.Namespace, Name: objRef.Name}, quota); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			// Add 'referred by a credentials binding' label
+			if !metav1.HasLabel(quota.ObjectMeta, v1beta1constants.LabelCredentialsBindingReference) {
+				patch := client.MergeFrom(quota.DeepCopy())
+				metav1.SetMetaDataLabel(&quota.ObjectMeta, v1beta1constants.LabelCredentialsBindingReference, "true")
+				if err := r.Client.Patch(ctx, quota, patch); err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to add referred label to the quota referenced in CredentialsBinding, quota: %s , namespace: %s : %w", quota.Name, quota.Namespace, err)
+				}
+			}
+		}
+	}
+
+	// Add 'referred by a credentials binding' label
+	if !metav1.HasLabel(secret.ObjectMeta, v1beta1constants.LabelCredentialsBindingReference) {
+		patch := client.MergeFrom(secret.DeepCopy())
+		metav1.SetMetaDataLabel(&secret.ObjectMeta, v1beta1constants.LabelCredentialsBindingReference, "true")
+		if err := r.Client.Patch(ctx, secret, patch); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add referred label to Secret referenced in CredentialsBinding: %w", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// We may only release a credential if there is no other credentials binding that references it (maybe in a different namespace).
+func (r *Reconciler) mayReleaseCredentials(ctx context.Context, binding *securityv1alpha1.CredentialsBinding) (bool, error) {
+	credentialsBindingList := &securityv1alpha1.CredentialsBindingList{}
+	if err := r.Client.List(ctx, credentialsBindingList); err != nil {
+		return false, err
+	}
+
+	for _, cb := range credentialsBindingList.Items {
+		// skip if it is one and the same credentials binding
+		if cb.Namespace == binding.Namespace && cb.Name == binding.Name {
+			continue
+		}
+		if cb.CredentialsRef.Namespace == binding.CredentialsRef.Namespace && cb.CredentialsRef.Name == binding.CredentialsRef.Name {
```

Comment:
The referred credential must be of the same APIVersion and Kind, you can have Secret and WorkloadIdentity having the same namespace and name, but they are different resources.

Reply:
Yes, nice catch